### PR TITLE
fix: skip manifest cache entries from a different generation policy (#29766)

### LINKED ISSUE #29766: Argo CD repo-server returns cached manifests after generator policy changes
Checklist:

- [x] I've included steps to reproduce the bug.
- [x] I've included the tested Argo CD version and source commit.
- [x] I've included the relevant output below and in the attached evidence.

**Describe the bug**

In Argo CD v3.5.3, repo-server's manifest cache key omits request values that
change manifest-generation semantics. A cache entry generated with permissive
Kustomize or Helm settings, or with one Config Management Plugin input path, is
returned after that policy or input path changes.

The fresh request uses the current policy or input path and produces an error
or a different manifest, while the ordinary cache-hit request returns the old
manifest.

**To Reproduce**

The attached runner executes the real `reposerver/repository` package and
checks three cases: Kustomize `LoadRestrictionsNone`, Helm allowed value-file
schemes, and CMP manifest-generation paths.

```sh
git clone --branch v3.5.3 --depth 1 https://github.com/argoproj/argo-cd.git argo-cd-v3.5.3
test "$(git -C argo-cd-v3.5.3 rev-parse HEAD)" = \
  c9c369efcc5b2a0bd720803f8d14a1c3eaddf579

ARGO_CD_DIR="$PWD/argo-cd-v3.5.3" \
KUSTOMIZE_BIN=/path/to/kustomize-v5.8.1 \
  ./attachments/reproduce.sh result
```

The complete command and dependency details are in `report.md` and
`attachments/README.md`.

**Expected behavior**

A cache hit should be used only when the cached request and the current request
have equivalent manifest-generation inputs and policy. A policy or input-path
change should cause a fresh generation or an invalidation.

**Screenshots**

Not applicable; this affects the repo-server repository service.

**Version**

```text
Argo CD v3.5.3
Source commit: c9c369efcc5b2a0bd720803f8d14a1c3eaddf579
Go: go1.23.5 linux/amd64
Kustomize: v5.8.1
```

**Logs**

The sanitized run is attached as `attachments/evidence/v3.5.3-run.log`. The
key observations are:

```text
cached-after-policy-change=[{"apiVersion":"v1","kind":"ConfigMap","metadata":{"name":"shared"}}]
fresh-strict-policy-error=... security ... file is not in or below ...
cached-after-scheme-removal=[{"apiVersion":"v1","kind":"ConfigMap","metadata":{"name":"remote-values-accepted"}}]
fresh-disallowed-scheme-error=... the URL scheme 'https' is not allowed
cached-after-path-change=[{"apiVersion":"v1","kind":"ConfigMap","metadata":{"name":"app-only"}}]
fresh-shared-root=[{"apiVersion":"v1","kind":"ConfigMap","metadata":{"name":"saw-shared-policy"}}]
```

**Attachments**

[detailed report.md](https://github.com/user-attachments/files/32336865/report.md)

[attachments.zip](https://github.com/user-attachments/files/32336866/attachments.zip)

### PR DESCRIPTION
**Pull request by hktitof**
Checklist:

* [x] Either (a) an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) was created and discussed with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes. (b, this is a bug fix)
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Title of the PR](https://argo-cd.readthedocs.io/en/latest/developer-guide/submit-your-pr/#title-of-the-pr)
* [x] "Fixes #29766" is included in the description to automatically close the associated issue.
* [x] Both the CLI and UI were checked for this change. (not affected, this is repo server internals only)
* [x] Does this PR require documentation updates? (no, internal correctness fix, the behavior now matches what users already expect when they change generation settings)
* [x] Documentation was updated as required by this PR. (nothing to update)
* [x] All commits are signed off as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] Unit and/or e2e tests were written for this change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] Does this PR follow the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines? (bug fix, no feature)
* [x] A brief description of why this PR is necessary and/or what this PR solves is included below.
* [ ] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, the older releases this fix should be cherry-picked into are indicated (this may or may not happen depending on risk/complexity). (release-3.0 and release-3.1, the change is small and additive)

## Why this PR is needed

Fixes #29766

the repo server manifest cache key covers revision, app source, ref sources, cluster runtime info and source integrity, but it ignores the request values that change what manifest generation is allowed to read and emit: KustomizeOptions, HelmOptions and the manifest-generate-paths annotation. so when an admin tightens those settings, an entry generated under the old permissive policy keeps getting served, which is exactly the three repro cases in the issue (kustomize load restrictions, helm value file schemes, cmp generate paths)

## What this PR does

- adds a generation policy hash field to CachedManifestResponse in reposerver/cache
- computes the hash from KustomizeOptions, HelmOptions and AnnotationManifestGeneratePaths of the request, a request without any of these values hashes to the empty string so old entries under the default policy still hit
- compares the hash on every cache hit in getManifestCacheEntry and treats a mismatch as a miss, so the fresh generation overwrites the stale entry, no key format change is needed
- also resets the failure-state entry in the error caching path when the policy differs, so a cached failure from another policy is not extended

the cache key format stays the same, so the revision rename path (updateCachedRevision) keeps matching and old entries miss exactly once and get rewritten

## Tests

- TestGenerateManifests_IgnoresCacheEntryFromDifferentGenerationPolicy fails on master (the stricter policy request gets served the cached response, same cache key mfst|...|2613152381|nil for both) and passes with this fix
- TestManifestGenerationPolicyHash covers the hash helper edge cases (nil, empty request, equal policies, each of the three values changing the hash)
- full go test on reposerver/cache and reposerver/repository: no new failures vs master (four helm related tests fail identically on pristine master and with this change in my sandbox, they need chart repo network access and two of them expect command strings from an older helm binary than the one in my sandbox)
- gofmt, go vet and golangci-lint clean on the touched packages

### CODE REVIEW

File: reposerver/cache/cache.go [hunk 1]
```diff
@@ -49,12 +50,55 @@ type ClusterRuntimeInfo interface {
...
+}
+
+// ManifestGenerationPolicyHash summarizes the manifest generation policy of a request. Cached
+// manifests are only valid while this value is unchanged, so a policy change results in a cache
+// miss and a fresh generation instead of a stale response.
+func ManifestGenerationPolicyHash(policy ManifestGenerationPolicy) string {
+	if policy == nil {
+		return ``
+	}
+	payload, err := json.Marshal(struct {
```
Comment (nitishfy) [inline reposerver/cache/cache.go]: Let's create a separate struct for this as it'd make the code more readable.

File: reposerver/cache/cache.go [hunk 2]
```diff
@@ -49,12 +50,55 @@ type ClusterRuntimeInfo interface {
...
 // CachedManifestResponse represents a cached result of a previous manifest generation operation, including the caching
 // of a manifest generation error, plus additional information on previous failures
 type CachedManifestResponse struct {
 	// NOTE: When adding fields to this struct, you MUST also update shallowCopy()
 
-	CacheEntryHash                  string                      `json:"cacheEntryHash"`
+	CacheEntryHash string `json:"cacheEntryHash"`
+	// GenerationPolicyHash summarizes the manifest generation policy this entry was created under
+	// (see ManifestGenerationPolicyHash). Entries stored before this field existed carry the empty
+	// value and are missed exactly once, after which the entry is rewritten under the current policy.
```
Comment (nitishfy) [inline reposerver/cache/cache.go:113]: From what i can see, `generateCacheEntryHash` marshals the whole struct  so old entries can have an extra hash content. WDYT?

File: reposerver/repository/repository.go [hunk 3]
```diff
@@ -1053,6 +1065,13 @@ func (s *Service) getManifestCacheEntry(revision string, q *apiclient.ManifestRe
 	res := cache.CachedManifestResponse{}
 	err := s.cache.GetManifests(cacheKey, &res)
 	if err == nil {
+		// The cached entry is only valid while the manifest generation policy that produced it is
+		// still in effect (Kustomize build options, Helm value file schemes, manifest-generate-paths).
+		if res.GenerationPolicyHash != cache.ManifestGenerationPolicyHash(q) {
+			logCtx.Info("manifest cache hit ignored: manifest generation policy changed since the entry was stored")
```
Comment (nitishfy) [inline reposerver/repository/repository.go]:
```suggestion
			logCtx.Debug("manifest cache hit ignored: manifest generation policy changed since the entry was stored")
```

File: reposerver/cache/cache_test.go [hunk 4]
```diff
@@ -355,6 +355,45 @@ func TestCachedManifestResponse_ShallowCopyExpectedFields(t *testing.T) {
...
+	assert.NotEqual(t,
+		ManifestGenerationPolicyHash(permissive),
+		ManifestGenerationPolicyHash(requestWithPolicy(``, []string{`https`}, ``)),
+		`different helm options produce different hashes`)
+	assert.NotEqual(t,
+		ManifestGenerationPolicyHash(permissive),
+		ManifestGenerationPolicyHash(requestWithPolicy(``, nil, `apps/*.yaml`)),
+		`different manifest generate paths produce different hashes`)
+}
+
```
Comment (nitishfy) [inline reposerver/cache/cache_test.go:401]: Can we also add a test which checks the same versions in a different order should give the same hash?

**Review by nitishfy — CHANGES_REQUESTED** [review]
Left some comments, PTAL.

File: reposerver/cache/cache.go (hunk 1 above)
Comment (hktitof) [reply reposerver/cache/cache.go]: done, pulled the anonymous struct into a named manifestGenerationPolicySnapshot type so the field order and json tags stay pinned in one place @nitishfy

File: reposerver/cache/cache.go (hunk 2 above)
Comment (hktitof) [reply reposerver/cache/cache.go:113]: i measured this, an entry stored before the field existed misses once bcs the entry hash covers it, gets rewritten under the current policy and after that the policy hash gates every hit, so old data never gets served under a different policy @nitishfy

File: reposerver/repository/repository.go (hunk 3 above)
Comment (hktitof) [reply reposerver/repository/repository.go]: applied, the log line is Debug now @nitishfy

File: reposerver/cache/cache_test.go (hunk 4 above)
Comment (hktitof) [reply reposerver/cache/cache_test.go:401]: added, the test checks the same value file schemes in a different order produce the same policy hash, and the hash now sorts the schemes first so the order genuinely can't matter @nitishfy
